File: src/spsc/tests/ProcessTree_Tests.scala
package spsc.tests

import org.junit.Test
import org.junit.Before
import org.junit.Assert._

class ProcessTree_Tests {

  var tree : Tree = null
  
  @Before def setup() : Unit = {
    tree = Tree.create(Var("r"))
    val r = tree.root
    tree = tree.addChildren(r,List((Var("m1"), null), (Var("m2"), null)))
    val m1 = tree.children(r)(0)
    val m2 = tree.children(r)(1)
    tree = tree.addChildren(m1, List((Var("n"), null)))
    tree = tree.replace(m2, Var("x"))
  }

  @Test def test01PrTreeBuilding() = {
    assertEquals(
    "{,0:(r,,,[1,4]),1:(m1,,0,[3]),3:(n,,1,[]),4:(x,,0,[])}",
    tree.toString)
  }

  def test02PrTreeLeaves() : Unit =
    assertEquals(List(3,2), tree.leaves map {n => n.nodeId})
}

File: spsc-lite-julia/test/process_tree_tests.jl
module ProcessTree_Tests

using Test
using SPSC.SLanguage
using SPSC.ProcessTree

@info "Testing ProcessTree"

@testset "Building process tree" begin

    t = Tree(Var("r")::Exp)
    r = t.root
    b1 = Branch(Var("m1"), nothing)
    b2 = Branch(Var("m2"), nothing)
    addChildren(t, r, Branch[b1, b2])
    m1 = r.children[1]
    m2 = r.children[2]
    addChildren(t, m1, [Branch(Var("n"), nothing)])
    replaceSubtree(t, m2, Var("x"))

    @test string(t) ==
        "{0:(r,nothing,nothing,[1,2]),1:(m1,nothing,0,[3]),3:(n,nothing,1,[]),2:(x,nothing,0,[])}"

    @test [n.nodeId for n in nodes(t)] == [0,1,3,2]

    @test [n.nodeId for n in leaves(t)] == [3,2]
end

end
